EIPS/eip-7620.md

```diff
@@ -0,0 +1,279 @@
+    
+This EIP uses terminology from the [EIP-3540](./eip-3540.md) which introduces the EOF format.
+
+EOF aims to remove code observability, which is a prerequisite to legacy EVM contract creation logic using create transactions, `CREATE` or `CREATE2`, because both the initcode and code are available to the EVM and can be manipulated. On the same premise, EOF removes opcodes like `CODECOPY` and `EXTCODECOPY`, introducing EOF subcontainers as a replacement to cater for factory contracts creating other contracts.
+    
+The new instructions and the new transaction type introduced in this EIP operate on EOF containers enabling all use cases of contract creation that legacy EVM has.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+    
+### Parameters
+
+| Constant | Value |
+| - | - |
+| `INITCODE_TX_TYPE` | `Bytes1(0x04)` |
+| `MAX_INITCODE_COUNT` | `256` |
+| `GKECCAK256WORD` | Defined as `6` in the Ethereum Yellow Paper |
+| `GCREATE` | Defined as `32000` in the Ethereum Yellow Paper |
+| `GCODEDEPOSIT` | Defined as `200` in the Ethereum Yellow Paper |
+| `GTXDATAZERO` | Defined as `4` in the Ethereum Yellow Paper |
+| `GTXDATANONZERO` | Defined as `16` in the Ethereum Yellow Paper |
+| `INITCODE_WORD_COST` | Defined as `2` in [EIP-3860](./eip-3860.md) |
+| `MAX_INITCODE_SIZE` | Defined as `2 * MAX_CODE_SIZE` in [EIP-3860](./eip-3860.md) |
+| `MAX_CODE_SIZE` | Defined as `24576` in [EIP-170](./eip-170.md) |
+
+We introduce three new instructions on the same block number [EIP-3540](./eip-3540.md) is activated on:
+
+1. `CREATE3` (`0xec`)
+2. `CREATE4` (`0xed`)
+3. `RETURNCONTRACT` (`0xee`)
+    
+### Transaction Types
+
+Introduce new transaction `InitcodeTransaction` (type `INITCODE_TX_TYPE`) which extends [EIP-1559](./eip-1559.md) (type 2) transaction by adding a new field `initcodes: List[ByteList[MAX_INITCODE_SIZE], MAX_INITCODE_COUNT]`.
+
+The `initcodes` can only be accessed via the `CREATE4` instruction (see below), therefore `InitcodeTransactions` are intended to be sent to contracts including `CREATE4` in their execution.
+
+We introduce a standardised Creator Contract (i.e. written in EVM, but existing at a known address, such as precompiles), which eliminates the need to have create transactions with empty `to`. Deployment of the Creator Contract will require an irregular state change at EOF activation block. Note that such introduction of the Creator Contract is needed, because only EOF contracts can create EOF contracts. See below for Creator Contract code.
+
+#### Gas schedule
+
+Each `initcodes` item data costs the same as calldata (`GTXDATANONZERO` gas for non-zero bytes, `GTXDATAZERO` for zero bytes -- see [EIP-2028](./eip-2028.md)). The intrinsic gas of an `InitcodeTransaction` is extended by the sum of all those items' costs.
+
```

Comment:
Perhaps some pseudocode (or actual code based on EELS) would be nice here to make the gas costs less ambiguous?